include/wx/msw/private/comstream.h

```diff
@@ -0,0 +1,179 @@
+    {
+        pstatstg->type = STGTY_STREAM;
+        pstatstg->cbSize.QuadPart = m_stream->GetSize();
+
+        return S_OK;
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE Clone(IStream **WXUNUSED(ppstm)) wxOVERRIDE
+    {
+        return E_NOTIMPL;
+    }
+
+    // ISequentialStream
+
+    virtual HRESULT STDMETHODCALLTYPE Read(void *WXUNUSED(pv),
+        ULONG WXUNUSED(cb), ULONG *WXUNUSED(pcbRead)) wxOVERRIDE
+    {
+        return E_NOTIMPL;
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE Write(const void *WXUNUSED(pv),
+        ULONG WXUNUSED(cb), ULONG *WXUNUSED(pcbWritten)) wxOVERRIDE
+    {
+        return E_NOTIMPL;
+    }
+
+
+protected:
+    wxStreamBase* m_stream;
+
+    static wxSeekMode OriginToSeekMode(DWORD origin)
+    {
+        switch (origin)
+        {
+        case STREAM_SEEK_SET:
+            return wxFromStart;
+        case STREAM_SEEK_CUR:
+            return wxFromCurrent;
+        case STREAM_SEEK_END:
+            return wxFromEnd;
+        }
+
+        return wxFromStart;
+    }
+};
+
+class wxCOMInputStreamAdapter : public wxCOMBaseStreamAdapter
+{
+public:
+    wxCOMInputStreamAdapter(wxInputStream* stream):
+        wxCOMBaseStreamAdapter(stream)
+    {
+
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE Seek(
+        LARGE_INTEGER dlibMove, DWORD dwOrigin,
+        ULARGE_INTEGER *plibNewPosition) wxOVERRIDE
+    {
+        wxFileOffset newOffset = reinterpret_cast<wxInputStream*>(m_stream)->SeekI(dlibMove.QuadPart, OriginToSeekMode(dwOrigin));
+        if (plibNewPosition)
+            plibNewPosition->QuadPart = newOffset;
+        return S_OK;
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE Read(void *pv, ULONG cb, ULONG *pcbRead) wxOVERRIDE
+    {
+        wxInputStream* inputStr = reinterpret_cast<wxInputStream*>(m_stream);
+        inputStr->Read(pv, cb);
+        *pcbRead = inputStr->LastRead();
+        return S_OK;
+    }
+
+};
+
+class wxCOMOutputStreamAdapter : public wxCOMBaseStreamAdapter
+{
+public:
+    wxCOMOutputStreamAdapter(wxOutputStream* stream) :
+        wxCOMBaseStreamAdapter(stream)
+    { }
+
+    virtual HRESULT STDMETHODCALLTYPE Seek(
+        LARGE_INTEGER dlibMove, DWORD dwOrigin,
+        ULARGE_INTEGER *plibNewPosition) wxOVERRIDE
+    {
+        wxFileOffset newOffset = reinterpret_cast<wxOutputStream*>(m_stream)->SeekO(dlibMove.QuadPart, OriginToSeekMode(dwOrigin));
+        if (plibNewPosition)
+            plibNewPosition->QuadPart = newOffset;
+        return S_OK;
+    }
+
+    virtual HRESULT STDMETHODCALLTYPE Write(const void *pv, ULONG cb, ULONG *pcbWritten) wxOVERRIDE
+    {
+        wxOutputStream* outputStr = reinterpret_cast<wxOutputStream*>(m_stream);
+        outputStr->Write(pv, cb);
+        *pcbWritten = outputStr->LastWrite();
+        return S_OK;
+    }
+
+};
+
+BEGIN_IID_TABLE(wxCOMBaseStreamAdapter)
+ADD_IID(Unknown)
+ADD_IID(SequentialStream)
+ADD_IID(Stream)
+END_IID_TABLE;
+
+IMPLEMENT_IUNKNOWN_METHODS(wxCOMBaseStreamAdapter);
```

Comment:
This is going to result in link errors if this header is included more than once, so it'd be better to move it into some existing file if possible.